daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala

```diff
@@ -33,11 +33,10 @@ object SResult {
       contractId: ContractId,
       templateId: TypeConName,
       committers: Set[Party],
-      // Callback to signal that the contract was not present
-      // or visible. Returns true if this was recoverable.
-      // TODO (MK) Drop now that tryHandleSubmitMustFail is dead.
-      cbMissing: Unit => Boolean,
-      cbPresent: ContractInst[Value.VersionedValue[ContractId]] => Unit,
+      // Callback
+      // returns the next expression to evaluate.
+      // In case of failure the call back does not throw but returns a SErrorDamlException
+      callback: ContractInst[Value.VersionedValue[ContractId]] => SExpr,
```

Comment:
I see two alternative: 
```suggestion
      // Callback
      // returns the next expression to evaluate as Right
      // In case of failure the call back does not throw but returns a Left[SErrorDamlException]
      callback: ContractInst[Value.VersionedValue[ContractId]] => Either[SErrorDamlException, SExpr],
```
Not sure it is very convince since `SErrorDamlException` is a subtype of `SExpr`.

```suggestion
      // Callback
      // In case of failure the callback set the `ctrl` of the machine that have returned this `SResult`  to a SErrorDamlException.
      callback: ContractInst[Value.VersionedValue[ContractId]] => Unit,
```
Not very convince..

Reply:
I don’t really get what you gain by moving from `Unit` to `SExpr`. The fact that the caller now has to mess with the speedy machine ctrl seems to somewhat mix up responsibilities. I think we’re better off in a state where the caller sets the `ctrl` once at the beginning and then everything else is handled by speedy itself (potentially by providing callback opaque to the caller).

So my suggestion would be to stick with `Unit` for the success case. I don’t care all that much if we use `Either` or just set the ctrl  for the error case so if you have a strong preference go for that.

Reply:
:+1:

Reply:
> I think we’re better off in a state where the caller sets the `ctrl` once at the beginning and then everything else is handled by speedy itself

I agree strongly with this.